File: python/liquid_crystal_print_task.py
#!/usr/bin/python
# coding: UTF-8


class LiquidCrystalPrintTask:
    def __init__(self):
        self.is_active = False  # このタスクが現在有効かどうか
        self.text = ""
        self.x = 0
        self.y = 0
        self.width = 0
        self.life_time = 0      # 表示する長さ（ミリ秒）。0なら永続
        self.registered_time = 0   # 表示を開始した時刻（ミリ秒）

        # エフェクト用の汎用設定変数枠
        self.effect_type = "NONE"  # "NONE", "BLINK", "SLIDE"
        self.delay_time = 0     # 表示開始前の待機時間（ミリ秒）。0なら待機なしで即表示
        self.effect_speed = 0
        self.effect_param = 0  # スライド方向（1:右、-1:左）などを格納

    def get_erase_data(self):
        return " " * self.width, self.x, self.y

    def evaluate(self, now):
        effect_start_time = self.registered_time + self.delay_time
        elapsed = now - effect_start_time

        if self.delay_time > 0 and elapsed < 0:
            # まだ待機時間中なので、窓の幅（width）と同じ長さの空白を返して、液晶には何も出さない
            return " " * self.width, self.x, self.y


        if self.life_time > 0 and elapsed >= self.life_time:
            self.is_active = False
            return self.get_erase_data()

        if self.effect_type == "NONE":
            return self.text, self.x, self.y

        # 点滅 (BLINK)
        elif self.effect_type == "BLINK":
            if (elapsed // self.effect_speed) % 2 == 0:
                return self.text, self.x, self.y
            else:
                return " " * len(self.text), self.x, self.y

        # スライド (SLIDE)
        elif self.effect_type == "SLIDE":
            frames = elapsed // self.effect_speed
            # current_text_x = self.x + (frames * self.effect_param)
            max_scroll = self.width + len(self.text)
            frames = frames % max_scroll
            if self.effect_param > 0:
                # ─── 右方向に流れるスライド ───
                # 左から空白を挿入し、窓の幅（width）に合わせて文字を切り出す
                raw_text = " " * frames + self.text
                text = raw_text[:self.width].ljust(self.width, " ")
            else:
                # ─── 左方向に流れるスライド ───
                raw_text = " " * self.width + self.text
                text = raw_text[frames : len(raw_text)].ljust(self.width, " ")
            return text, self.x, self.y
        # ナイトライダー (KNIGHT)
        elif self.effect_type == "KNIGHT":
            frames = elapsed // self.effect_speed
            max_spaces = self.width - len(self.text)
            if max_spaces <= 0:
                return self.text, self.x, self.y

            cycle = max_spaces * 2
            current_step = frames % cycle

            # ─── 【追加】右スタートのロジック ───
            # effect_param が負（例: -1）なら、時間を半分（片道分）進めた状態から始める
            if self.effect_param < 0:
                current_step = (current_step + max_spaces) % cycle

            # ─── 往復の空白数の計算 ───
            if current_step <= max_spaces:
                spaces_count = current_step
            else:
                spaces_count = cycle - current_step

            text = (" " * spaces_count + self.text).ljust(self.width, " ")
            return text, self.x, self.y

        return self.text, self.x, self.y

File: cpp/LiquidCrystalPrintTask.cpp
#include "LiquidCrystalPrintTask.h"
#include <string.h>
#include <stdio.h>
#include <stdlib.h>

LiquidCrystalPrintTask::LiquidCrystalPrintTask() {
    is_active = false;
    text[0] = '\0';
    x = 0;
    y = 0;
    width = 0;
    life_time = 0;
    registered_time = 0;
    effect_type = EffectType::NONE;
    delay_time = 0;
    effect_speed = 0;
    effect_param = 0;
}

EvaluateResult LiquidCrystalPrintTask::get_erase_data() const {
    EvaluateResult result;
    result.x = x;
    result.y = y;
    
    uint8_t len = (width < sizeof(result.text) - 1) ? width : sizeof(result.text) - 1;
    memset(result.text, ' ', len);
    result.text[len] = '\0';
    
    return result;
}

EvaluateResult LiquidCrystalPrintTask::evaluate(uint32_t now) {
    EvaluateResult result;
    result.x = x;
    result.y = y;

    uint32_t effect_start_time = registered_time + delay_time;
    int32_t elapsed = (int32_t)(now - effect_start_time);

    if (delay_time > 0 && elapsed < 0) {
        return get_erase_data();
    }

    if (life_time > 0 && (uint32_t)elapsed >= life_time) {
        is_active = false;
        return get_erase_data();
    }

    uint16_t text_len = strlen(text);

    if (effect_type == EffectType::NONE) {
        snprintf(result.text, sizeof(result.text), "%s", text);
        return result;
    }

    else if (effect_type == EffectType::BLINK) {
        if (effect_speed == 0 || (elapsed / effect_speed) % 2 == 0) {
            snprintf(result.text, sizeof(result.text), "%s", text);
        } else {
            memset(result.text, ' ', text_len);
            result.text[text_len] = '\0';
        }
        return result;
    }

    else if (effect_type == EffectType::SLIDE) {
        if (effect_speed == 0) {
            snprintf(result.text, sizeof(result.text), "%s", text);
            return result;
        }
        uint32_t frames = (elapsed / effect_speed) % (width + text_len);

        memset(result.text, ' ', width);
        result.text[width] = '\0';

        if (effect_param > 0) {
            for (uint16_t i = 0; i < text_len; ++i) {
                uint16_t tx = frames + i;
                if (tx < width) result.text[tx] = text[i];
            }
        } else {
            int16_t start_idx = frames - width;
            for (uint16_t i = 0; i < width; ++i) {
                int16_t text_pos = start_idx + i;
                if (text_pos >= 0 && text_pos < text_len) result.text[i] = text[text_pos];
            }
        }
        return result;
    }

    else if (effect_type == EffectType::KNIGHT) {
        int16_t max_spaces = width - text_len;
        if (max_spaces <= 0 || effect_speed == 0) {
            snprintf(result.text, sizeof(result.text), "%s", text);
            return result;
        }

        uint32_t frames = elapsed / effect_speed;
        uint16_t cycle = max_spaces * 2;
        uint16_t current_step = frames % cycle;

        if (effect_param < 0) {
            current_step = (current_step + max_spaces) % cycle;
        }

        uint16_t spaces_count = (current_step <= max_spaces) ? current_step : (cycle - current_step);

        memset(result.text, ' ', width);
        result.text[width] = '\0';
        for (uint16_t i = 0; i < text_len; ++i) {
            if (spaces_count + i < width) result.text[spaces_count + i] = text[i];
        }
        return result;
    }

    snprintf(result.text, sizeof(result.text), "%s", text);
    return result;
}












